keps/sig-apps/sidecarcontainers.md

```diff
@@ -111,21 +112,45 @@ This will be useful in scenarios such as when your sidecar is a proxy so that it
 
 To solve the problem of Jobs that don't complete: When RestartPolicy!=Always if all normal containers have reached a terminal state (Succeeded for restartPolicy=OnFailure, or Succeeded/Failed for restartPolicy=Never), then all sidecar containers will be sent a SIGTERM.
 
+Sidecars are just normal containers in almost all respects, they have all the same attributes, they are included in pod state, obey pod restart policy etc. The only differences are lifecycle related.
+
 ### Implementation Details/Notes/Constraints
 
-As this is a fairly large change I think it make sense to break this proposal down and phase in more functionality as we go, potential roadmap could look like:
+The proposal can broken down into four key pieces of implementation that all relatively separate from one another:
 
-* Add sidecar field, use it for the shutdown triggering when RestartPolicy!=Always
+* Shutdown triggering for sidecars when RestartPolicy!=Always
 * Pre-stop hooks sent to sidecars before non sidecar containers
 * Sidecars are terminated after normal containers
 * Sidecars start before normal containers
 
-
 As this is a change to the Container spec we will be using feature gating, you will be required to explicitly enable this feature on the api server as recommended [here](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions).
 
+#### Kubelet Changes:
+Broad outline of what places could be modified to implement desired behaviour:
+
+###### Shutdown triggering:
+Package `kuberuntime`
+
+Modify `kuberuntime_manager.go`, function `computePodActions`. Have a check in this function that will see if all the non-sidecars had permanently exited, if true: return all the running sidecars in `ContainersToKill`.
```

Comment:
When we trigger shutdown of sidecars through `ContainersToKill`, what exit status do they get? Do we send them SIGTERM and give them a chance to terminate gracefully? We need to verify that the overall Pod phase (Succeeded/Failed) will reflect the right semantics in the presence of sidecars, so that Job will work as intended.

Reply:
So when you return `ContainersToKill` it iterates over them and does `KillContainer` on each container. This is the function that sends prestop, sigterms/kills etc, grace period is obeyed. So as long as the sidecar exits `0` it should count towards succeeded.